### PR TITLE
KEP-4210: add support for ImageMaximumGCAge field

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
adds support for ImageGCMaximumAge kubelet configuration field, which sets the maximum age of an image before it's deemed unused and worthy of garbage collecting.
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Add ImageMaximumGCAge field to Kubelet configuration, which allows a user to set the maximum age an image is unused before it's garbage collected.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/apis/config/types.go
Comment (kannon92) [inline pkg/kubelet/apis/config/types.go:200]: Not sure if this is WIP but I would expect to see someplace where we default here.

Do you need to add some code in apis/config/v1beta1 to default this to 0?

File: pkg/kubelet/metrics/metrics.go
Comment (kannon92) [inline pkg/kubelet/metrics/metrics.go:793]: In other areas of the code, I was told that we don't add metrics for alpha features.  

I would like to include this in alpha but not sure if this is usually what we do for alpha features.

Comment (haircommander) [reply pkg/kubelet/metrics/metrics.go:793]: this isn't so much a metric for an alpha feature, but a metric that will be useful to inform how well an alpha feature is working compared to when the feature is turned off. it'll measure GCs independent of the feature being on

**Review by kannon92 — COMMENTED** [review]
I think you need this guarded by a feature toggle.

In the KEP you mention that you are adding a feature toggle for this: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4210-max-image-gc-age/kep.yaml

**Comment by haircommander** [comment]
> I think you need this guarded by a feature toggle.

great catch! totally forgot about that--I've added it.

File: pkg/kubelet/kubelet.go
Comment (kannon92) [inline pkg/kubelet/kubelet.go:429]: I would probably check for nil in addtion to zero.  

`ImageMaximumGCAge` is omitEmpty so I think it could be empty.

File: pkg/kubelet/metrics/metrics.go
Comment (kannon92) [reply pkg/kubelet/metrics/metrics.go:793]: That sounds good to me.

File: pkg/kubelet/apis/config/fuzzer/fuzzer.go
Comment (kannon92) [inline pkg/kubelet/apis/config/fuzzer/fuzzer.go:64]: Should you just use the default of 0s here?

File: pkg/kubelet/kubelet.go
Comment (kannon92) [reply pkg/kubelet/kubelet.go:429]: Hmm... So the staging api says this is omity empty but the kubelet/apis/config/type.go is not..

Not sure if this would actually be nil..

If Kubelet was upgraded and restarted, I assume it reads the config first and defaults this to zero?

File: pkg/kubelet/kubelet.go
Comment (kannon92) [inline pkg/kubelet/kubelet.go:427]: Could you look into adding a unit test for this?  Kubelet has a corresponding unit test in this file so if possible it would be nice to verify the behavior of the feature toggle.

File: pkg/kubelet/images/image_gc_manager.go
Comment (kannon92) [inline pkg/kubelet/images/image_gc_manager.go:352]:
```suggestion
		klog.V(5).InfoS("Evaluating image ID for possible garbage collection based on image age", "imageID", image.id)
```

File: pkg/kubelet/images/image_gc_manager.go
Comment (kannon92) [inline pkg/kubelet/images/image_gc_manager.go:83]: nit: it would be nice to maybe mention that this is an alpha feature. If feature is disabled, this is set to 0.

File: pkg/kubelet/images/image_gc_manager.go
Comment (kannon92) [reply pkg/kubelet/images/image_gc_manager.go:352]: also drop the comment.

File: pkg/kubelet/kubelet.go
Comment (haircommander) [reply pkg/kubelet/kubelet.go:429]: > I would probably check for nil in addtion to zero.

atually ImageMaximumGCAge is a [struct](https://github.com/kubernetes/kubernetes/pull/121275/files#diff-8a6475c76df7985730ad410e5d1a6caccbbe3ba44853ed773722063cf3625639R200) so a nil check fails

> Hmm... So the staging api says this is omity empty but the kubelet/apis/config/type.go is not..

I followed ImageMinimumGCAge here, not sure why it's omit empty on one but not the other

File: pkg/kubelet/apis/config/fuzzer/fuzzer.go
Comment (haircommander) [reply pkg/kubelet/apis/config/fuzzer/fuzzer.go:64]: I believe that is what is done by this. Duration is just a int64:
```
type Duration int64
```

File: pkg/kubelet/kubelet.go
Comment (haircommander) [reply pkg/kubelet/kubelet.go:427]: what are you hoping to be tested for @kannon92 ? the MaxAge variable in the GC manager is set?

Comment (kannon92) [reply pkg/kubelet/kubelet.go:427]: Yea, I normally would like to make sure that if someone turns off the feature does it revert to old behavior.  And if its turned on, does it work as intended?

It's not a big deal though.

Comment (kannon92) [reply pkg/kubelet/kubelet.go:427]: SGTM.

File: pkg/kubelet/images/image_gc_manager_test.go
Comment (kannon92) [inline pkg/kubelet/images/image_gc_manager_test.go:661]: Could you add a test with `MaxAge` being set to 0?

After that, I think this LGTM.

Comment (haircommander) [reply pkg/kubelet/images/image_gc_manager_test.go:661]: updated!

**Comment by kannon92** [comment]
LGTM.  I'll leave the label for a sig-node reviewer/approver.  

/assign @rphillips @mrunalp @dchen1107

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:288]: revert this change, docs use the serialized field name, so lowercase

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:293]: start the doc with a field name matching the serialized field name

File: pkg/kubelet/apis/config/types.go
Comment (liggitt) [inline pkg/kubelet/apis/config/types.go:199]:
```suggestion
	// The default of this field is "0s", which disables it.
```

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:295]:
```suggestion
	// Default: "0s"
```

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:294]:
```suggestion
	// The default of this field is "0s", which disables it.
```

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:297]: add validation for this field to the kubelet config validation function:
* must not be negative
* must not be < ImageMinimumGCAge if non-zero?
* anything else?

Comment (haircommander) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:296]: I also added a check for the featuregate

File: pkg/kubelet/apis/config/types.go
Comment (liggitt) [inline pkg/kubelet/apis/config/types.go:199]: I guess "which disabled it" could be read two ways – does it disable GC or mean there is no max age? make that clearer in the doc (here and in v1beta1/types.go)

**Review by liggitt — COMMENTED** [review]
API bits lgtm. One nit on the field doc.

Once this has lgtm / approve for non API packages tag me and I'll add approve for the API config field package

**Comment by kannon92** [comment]
/lgtm

@assign @dchen1107 @mrunalp

**Review by mrunalp — APPROVED** [review]

**Pull request merged**
